libirods_rule_engine_plugin-indexing.cpp

```diff
@@ -420,75 +579,189 @@ namespace {
+            {".xul", "application/vnd.mozilla.xul+xml"},
+            {".zip", "application/zip"},
+            {".3gp", "video/3gpp"}, // audio/3gpp if it doesn't contain video,
+            {".3g2", "video/3gpp2"}, // audio/3gpp2 if it doesn't contain video,
+            {".7z", "application/x-7z-compressed"},
+        };
+        std::string retvalue {};
+        if (auto offs = input.rfind("."); offs != std::string::npos) {
+            retvalue = default_mimeTypes[ to_lowercase(input.substr(offs)) ];
+        }
+        return retvalue.size() ? retvalue : "application/octet-stream";
+    }
+
+    auto get_system_metadata( ruleExecInfo_t* _rei
+                             ,const std::string& _obj_path ) -> nlohmann::json {
+        using nlohmann::json;
+        const boost::filesystem::path p{_obj_path};
+        const std::string parent_name = p.parent_path().string();
+        const std::string name = p.filename().string();
+        namespace fs   = irods::experimental::filesystem;
+        namespace fsvr = irods::experimental::filesystem::server;
+        auto irods_path = fs::path{_obj_path};
+        const auto s = fsvr::status(*_rei->rsComm, irods_path);
+        std::string query_str;
+
+        json obj;
+
+        obj["absolutePath"] = _obj_path;
+
+        if (fsvr::is_data_object(s)) {
+            query_str = fmt::format("SELECT DATA_ID , DATA_MODIFY_TIME, DATA_ZONE_NAME, COLL_NAME, DATA_SIZE where DATA_NAME = '{0}'"
+                                    " and COLL_NAME = '{1}' ", name, parent_name  );
+            irods::query<rsComm_t> qobj{_rei->rsComm, query_str, 1};
+            for (const auto & i:qobj) {
+                obj["lastModifiedDate"] = std::stol( i[1] ) * 1000; // epoch ms
+                obj["zoneName"] = i[2];
+                obj["parentPath"] = i[3];
+                obj["dataSize"] = std::stol( i[4] );
+                obj["isFile"] = true;
+                break;
+            }
+        }
+        else if (fsvr::is_collection(s)) {
+            query_str = fmt::format("SELECT COLL_ID , COLL_MODIFY_TIME, COLL_ZONE_NAME, COLL_PARENT_NAME where COLL_NAME = '{0}'"
+                                    " and COLL_PARENT_NAME = '{1}' ", _obj_path, parent_name  );
+            irods::query<rsComm_t> qobj{_rei->rsComm, query_str, 1};
+            for (const auto & i : qobj) {
+                obj["lastModifiedDate"] = std::stol( i[1] ) * 1000; // epoch ms
+                obj["zoneName"] = i[2];
+                obj["parentPath"] = i[3];
+                obj["dataSize"] = 0L;
+                obj["isFile"] = false;
+                break;
+            }
+        }
+        auto fileName = obj ["fileName"] = irods_path.object_name();
+        obj ["url"] = fmt::format(config->urlTemplate, _obj_path);
+        obj["mimeType"] = get_default_Mime_Type (fileName);  // dwm - Q for Mike : what is mimetype for collections ?
```

Comment:
@michael-conway : What's the mime type for a collection ?

Reply:
@trel - does this still matter ?  @michael-conway does your Java indexer code populate the mime type for a collection object?

Reply:
unix/windows/linux don't agree, and most don't have consistent answers or any at all.

I don't think collections should have a mime-type.

Reply:
In that case, I wonder if  the lack-of-mimeType should be represented by an empty string, or  a **null** object ?

@michael-conway  what does Metalnx and/or  NIEHS metadata formal schema "expect"?  
AFAIK,  the  entry in the index will have to have a mimeType key whether it concerns a collection or data object ... Basically, does Metalnx have an opinion what it should be for a "directory" / "collection" ?

For what it's worth I never had success running the example indexer (https://github.com/trel/metadata-and-file-props-indexer) , but could try again. I don't know if I should be following its example or whether there's a formal specification.